include/message_filters/sync_policies/approximate_epsilon_time.h

```diff
@@ -0,0 +1,357 @@
+  }
+
+  template <size_t Is>
+  bool
+  check_timestamp_within_epsilon(const TimeIndexPair & older)
+  {
+    namespace mt = message_filters::message_traits;
+    using ThisEventType = typename std::tuple_element<Is, Events>::type;
+    if constexpr (Is >= RealTypeCount::value) {
+      return true;
+    }
+    if (Is == older.second) {
+      return true;
+    }
+    const auto & events_of_this_type = std::get<Is>(events_);
+    if (0u == events_of_this_type.size()) {
+      // this condition should not happen
+      return false;
+    }
+    auto ts = mt::TimeStamp<typename ThisEventType::Message>::value(*events_of_this_type.at(0).getMessage());
+    if (older.first + epsilon_ >= ts) {
+      return true;
+    }
+    return false;
+  }
+
+  template <size_t ... Is>
+  bool
+  check_all_timestamp_within_epsilon_helper(
+    const TimeIndexPair & older, std::index_sequence<Is...> const &)
+  {
+    bool valid = true;
+    ((valid &= check_timestamp_within_epsilon<Is>(older)), ...);
+    return valid;
+  }
+
+  bool
+  check_all_timestamp_within_epsilon(const TimeIndexPair & older)
+  {
+    return check_all_timestamp_within_epsilon_helper(
+      older, std::make_index_sequence<9u>());
+  }
+
+  // void erase_old_events()
+  // {
+  //   auto old = get_older_timestamp();
+  //   erase_old_events_within_ts(old);
+  // }
+
+  template<size_t Is>
+  void
+  erase_begin_of_vector()
+  {
+    if constexpr (Is >= RealTypeCount::value) {
+      return;
+    }
+    auto & this_vector = std::get<Is>(events_);
+    if (this_vector.begin() != this_vector.end()) {
+      this_vector.erase(this_vector.begin());
+      if (this_vector.empty()) {
+        --number_of_non_empty_events_;
+      }
+    }
+  }
+
+  template <size_t ... Is>
+  void
+  erase_begin_of_vectors_helper(std::index_sequence<Is...> const &)
+  {
+    ((erase_begin_of_vector<Is>()), ...);
+  }
+
+  void erase_begin_of_vectors()
+  {
+    return erase_begin_of_vectors_helper(std::make_index_sequence<9u>());
+  }
+
+  template<size_t Is>
+  void
+  erase_begin_of_vector_if_within_ts(rclcpp::Time timestamp)
+  {
+    namespace mt = message_filters::message_traits;
+    using ThisEventType = typename std::tuple_element<Is, Events>::type;
+    if constexpr (Is >= RealTypeCount::value) {
+      return;
+    }
+    auto & this_vector = std::get<Is>(events_);
+    if (this_vector.begin() == this_vector.end()) {
+      return;
+    }
+    auto event_ts = mt::TimeStamp<typename ThisEventType::Message>::value(
+      *this_vector.at(0).getMessage());
+    if (timestamp + epsilon_ < event_ts) {
+      return;
+    }
+    this_vector.erase(this_vector.begin());
+    if (this_vector.empty()) {
+      --number_of_non_empty_events_;
+    }
+  }
+
+  template <size_t ... Is>
+  void
+  erase_old_events_within_ts_helper(rclcpp::Time timestamp, std::index_sequence<Is...> const &)
+  {
+    ((erase_begin_of_vector_if_within_ts<Is>(timestamp)), ...);
```

Comment:
```suggestion
    ((erase_beginning_of_vector_if_earlier_than_ts<Is>(timestamp)), ...);
```

Reply:
see https://github.com/ros2/message_filters/pull/84#discussion_r967390918

Reply:
Fixed in https://github.com/ros2/message_filters/commit/41d66703a2679d72bfbf85e4c87fb74d1138de4e